### ISSUE TITLE
`batch_size`>1 is not working in `train_epoch` function

### ISSUE THREAD

**Issue by patel-zeel** [tom-andersson/deepsensor#32]
Hi,

I was trying to run the following README example

```py
import deepsensor.torch
from deepsensor.data.loader import TaskLoader
from deepsensor.data.processor import DataProcessor
from deepsensor.model.convnp import ConvNP
from deepsensor.train.train import train_epoch

import xarray as xr
import pandas as pd
import numpy as np

# Load raw data
ds_raw = xr.tutorial.open_dataset("air_temperature")

# Normalise data
data_processor = DataProcessor(x1_name="lat", x1_map=(15, 75), x2_name="lon", x2_map=(200, 330))
ds = data_processor(ds_raw)

# Set up task loader
task_loader = TaskLoader(context=ds, target=ds)

# Set up model
model = ConvNP(data_processor, task_loader)

# Generate training tasks with up to 10% of grid cells passed as context and all grid cells
# passed as targets
train_tasks = []
for date in pd.date_range("2013-01-01", "2014-11-30")[::7]:
    task = task_loader(date, context_sampling=np.random.uniform(0.0, 0.1), target_sampling="all")
    train_tasks.append(task)

print(len(train_tasks))
# Train model
for epoch in range(10):
    train_epoch(model, train_tasks, progress_bar=True, batch_size=2)

# Predict on new task with 10% of context data and a dense grid of target points
test_task = task_loader("2014-12-31", 0.1)
mean_ds, std_ds = model.predict(test_task, X_t=ds_raw)
```

When I set `batch_size` greater than 1, I get the following error.
```
---------------------------------------------------------------------------
NotFoundLookupError                       Traceback (most recent call last)
Cell In[11], line 34
     32 # Train model
     33 for epoch in range(10):
---> 34     train_epoch(model, train_tasks, progress_bar=True, batch_size=2)
     36 # Predict on new task with 10% of context data and a dense grid of target points
     37 test_task = task_loader("2014-12-31", 0.1)

File [~/deepsensor/deepsensor/train/train.py:115](https://vscode-remote+ssh-002dremote-002b7b22686f73744e616d65223a2252616d616e756a616e227d.vscode-resource.vscode-cdn.net/home/patel_zeel/deepsensor/dist/~/deepsensor/deepsensor/train/train.py:115), in train_epoch(model, tasks, lr, batch_size, opt, progress_bar, tqdm_notebook)
    113 for batch_i in tqdm(range(n_batches), disable=not progress_bar):
    114     if batch_size is not None:
--> 115         task = concat_tasks(
    116             tasks[batch_i * batch_size : (batch_i + 1) * batch_size]
    117         )
    118     else:
    119         task = tasks[batch_i]

File [~/deepsensor/deepsensor/model/convnp.py:546](https://vscode-remote+ssh-002dremote-002b7b22686f73744e616d65223a2252616d616e756a616e227d.vscode-resource.vscode-cdn.net/home/patel_zeel/deepsensor/dist/~/deepsensor/deepsensor/model/convnp.py:546), in concat_tasks(tasks, multiple)
    543 merged_task["Y_c"] = [c[1] for c in merged_context]
    545 # This assumes that all tasks have the same number of targets
--> 546 merged_task["X_t"] = [
    547     B.concat(*[t["X_t"][i] for t in tasks], axis=0)
    548     for i in range(len(tasks[0]["X_t"]))
    549 ]
    550 merged_task["Y_t"] = [
    551     B.concat(*[t["Y_t"][i] for t in tasks], axis=0)
    552     for i in range(len(tasks[0]["Y_t"]))
    553 ]
    554 # # Multiple target sets, different target locations
    555 # xts = [task["X_t"] for task in tasks]
    556 # yts = [task["Y_t"] for task in tasks]
    557 # merged_task["X_t"] = backend.nps.AggregateInput(*[(xt, i) for i, xt in enumerate(xts)])
    558 # merged_task["Y_t"] = backend.nps.Aggregate(*yts)

File [~/deepsensor/deepsensor/model/convnp.py:547](https://vscode-remote+ssh-002dremote-002b7b22686f73744e616d65223a2252616d616e756a616e227d.vscode-resource.vscode-cdn.net/home/patel_zeel/deepsensor/dist/~/deepsensor/deepsensor/model/convnp.py:547), in (.0)
    543 merged_task["Y_c"] = [c[1] for c in merged_context]
    545 # This assumes that all tasks have the same number of targets
    546 merged_task["X_t"] = [
--> 547     B.concat(*[t["X_t"][i] for t in tasks], axis=0)
    548     for i in range(len(tasks[0]["X_t"]))
    549 ]
    550 merged_task["Y_t"] = [
    551     B.concat(*[t["Y_t"][i] for t in tasks], axis=0)
    552     for i in range(len(tasks[0]["Y_t"]))
    553 ]
    554 # # Multiple target sets, different target locations
    555 # xts = [task["X_t"] for task in tasks]
    556 # yts = [task["Y_t"] for task in tasks]
    557 # merged_task["X_t"] = backend.nps.AggregateInput(*[(xt, i) for i, xt in enumerate(xts)])
    558 # merged_task["Y_t"] = backend.nps.Aggregate(*yts)

File [~/miniconda3/envs/active_learn/lib/python3.10/site-packages/plum/function.py:392](https://vscode-remote+ssh-002dremote-002b7b22686f73744e616d65223a2252616d616e756a616e227d.vscode-resource.vscode-cdn.net/home/patel_zeel/deepsensor/dist/~/miniconda3/envs/active_learn/lib/python3.10/site-packages/plum/function.py:392), in Function.__call__(self, *args, **kw_args)
    388 except KeyError:
    389     # Cache miss. Run the resolver based on the arguments.
    390     method, return_type = self.resolve_method(args, types)
--> 392 return _convert(method(*args, **kw_args), return_type)

File [~/miniconda3/envs/active_learn/lib/python3.10/site-packages/lab/util.py:216](https://vscode-remote+ssh-002dremote-002b7b22686f73744e616d65223a2252616d616e756a616e227d.vscode-resource.vscode-cdn.net/home/patel_zeel/deepsensor/dist/~/miniconda3/envs/active_learn/lib/python3.10/site-packages/lab/util.py:216), in abstract..decorator..wrapper(*args, **kw_args)
    214     signature = plum.signature.Signature(*types_after)
    215     # TODO: Use the message from Plum directly here.
--> 216     raise plum.NotFoundLookupError(
    217         f'For function "{f.__name__}", signature {signature} could not be '
    218         f"resolved."
    219     )
    221 # Retry call.
    222 return getattr(B, f.__name__)(*args, **kw_args)

NotFoundLookupError: For function "concat", signature Signature(tuple, tuple) could not be resolved.
```

**Comment by tom-andersson** [comment]
Hey @patel-zeel, thanks for trying out DeepSensor and for catching and raising this bug. The issue was that when `target_sampling="all"` (in the `TaskLoader.__call__`) for sampling a `target` dataset that is a gridded `xarray` object, this produces a tuple of coordinates like `(x1, x2)` in that `Task["X_t"]` entry. Concatenation of tuples of coords for batching `Task`s was not supported in `deepsensor.convnp.concat_tasks` (which is what's called by `train_epoch` when `batch_size > 1`.

This is now fixed in `main` by https://github.com/tom-andersson/deepsensor/commit/1aac5f0702cb4688f923c569faa0b41d76aff0b6

**Comment by tom-andersson** [comment]
On a related note, an error is raised when merging context sets if one of those context sets contains zero observations. Currently, this can happen in the above quickstart code because the number of context points is a random fraction in `[0, 0.1]`. 

I believe this is an issue on the `neuralprocesses` side so I've raised an issue over there: https://github.com/wesselb/neuralprocesses/issues/16

**Comment by patel-zeel** [comment]
Perfect, @tom-andersson! Thanks a lot for the detailed information and the fix! I must say `deepsensor` has an easy-to-use interface similar to `sklearn` and its seamless integration with `xarray` saves a lot of time and energy to preprocess and post-process the data.

**Comment by tom-andersson** [comment]
Thank you for the kind words @patel-zeel!

**Comment by tom-andersson** [comment]
The concatting-zero-contexts bug mentioned above has now been fixed by @wesselb in v0.2.2 of `neuralprocesses` (https://github.com/wesselb/neuralprocesses/releases/tag/v0.2.2).
